File: DailyChallenges/CPP/LeetCode_328.cpp
// 328. Odd Even Linked List @ 2022/12/06

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return head;

        ListNode* evenHead = head->next, *oddTail = head, *evenTail = evenHead, *next = evenTail->next;
        bool isOdd = true;
        while (next){
            if (isOdd)
                oddTail->next = next, oddTail = next;
            else
                evenTail->next = next, evenTail = next;
            next = next->next;
            isOdd = !isOdd;
        }

        oddTail->next = evenHead, evenTail->next = nullptr;
        return head;
    }
};

File: DailyChallenges/Scala/LeetCode_328.scala
object Solution {
    def oddEvenList(head: ListNode): ListNode = {
        if (head == null || head.next == null) head else {
            @scala.annotation.tailrec
            def helper(node: ListNode, isOdd: Boolean, oddTail: ListNode, evenTail: ListNode): ListNode = {
                if (node == null){
                    evenTail.next = null
                    oddTail
                }else{
                    if (isOdd) oddTail.next = node else evenTail.next = node
                    helper(node.next, !isOdd, if (isOdd) node else oddTail, if (!isOdd) node else evenTail)
                }
            }

            val evenHead = head.next
            val oddTail = helper(evenHead.next, true, head, evenHead)
            oddTail.next = evenHead
            head
        }
    }
}
